src/database.js

```diff
@@ -205,126 +234,215 @@ class Database {
+    ): Promise<TestSummary> {
+        return this.cannotCommit(auth, [Batch.set(path, data)]);
     }
 
     async canUpdate(
         auth: FirestoreAuth,
         path: string,
         data: Object
-    ): Promise<boolean> {
-        const result = await this.testRules(
-            createUpdateTest(true, auth, path, data)
-        );
-
-        return result.state == 'SUCCESS';
+    ): Promise<TestSummary> {
+        return this.canCommit(auth, [Batch.update(path, data)]);
     }
 
     async cannotUpdate(
         auth: FirestoreAuth,
         path: string,
         data: Object
-    ): Promise<boolean> {
-        const result = await this.testRules(
-            createUpdateTest(false, auth, path, data)
+    ): Promise<TestSummary> {
+        return this.cannotCommit(auth, [Batch.update(path, data)]);
+    }
+
+    /*
+     * Factories for tests
+     */
+    createGetTest(
+        allow: boolean,
+        auth: FirestoreAuth,
+        path: string
+    ): FirestoreTestCase {
+        const functionMocks = this.createMockFunctions();
+        const doc = this.getDocument(path);
+        const request = {
+            auth,
+            path: createDocumentPath(path),
+            method: 'get'
+        };
+
+        return {
+            expectation: allow ? 'ALLOW' : 'DENY',
+            request,
+            resource: { data: doc ? doc.fields : null },
+            functionMocks
+        };
+    }
+
+    /*
+     * Create a test for a commit with multiple updates.
+     */
+    createCommitTest(
+        allow: boolean,
+        auth: FirestoreAuth,
+        batch: BatchOperation[]
+    ): FirestoreTestCase[] {
+        const expectation = allow ? 'ALLOW' : 'DENY';
+        const baseFunctionMocks = this.createMockFunctions();
+        const afterFunctionMocks = this.createBatchAfterFunctionMocks(batch);
+
+        const functionMocks = [...baseFunctionMocks, ...afterFunctionMocks];
+
+        return batch.map(operation => {
+            const doc = this.getDocument(operation.document);
+            const request = {
+                auth,
+                path: createDocumentPath(operation.document),
+                method:
+                    operation.method == 'set'
+                        ? doc ? 'update' : 'create'
+                        : operation.method
+            };
+            const resource = {
+                data: operation.data || null
+            };
+
+            return {
+                expectation,
+                request,
+                resource,
+                functionMocks
+            };
+        });
+    }
+
+    /*
+     * Create the mocks for the API to represent the dataset.
+     */
+    createMockFunctions(): FirestoreMockFunction[] {
+        const documents = this.getDocuments();
+        const defaults = [
+            {
+                function: 'get',
+                args: [{ anyValue: {} }],
+                result: {
+                    value: null
```

Comment:
@samtstern Can you confirm that when a document doesn't exist, `get` and `getAfter` return `null` ?

Reply:
Yes that's correct, both return a `null` resource object.